File: src/main/java/com/star/easydoc/listener/AppActivationListener.java
package com.star.easydoc.listener;

import java.awt.*;
import java.net.URI;

import com.intellij.notification.Notification;
import com.intellij.notification.NotificationAction;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.application.ApplicationActivationListener;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.wm.IdeFrame;
import com.star.easydoc.common.util.NotificationUtil;
import com.star.easydoc.config.EasyDocConfig;
import com.star.easydoc.config.EasyDocConfigComponent;
import com.star.easydoc.service.gpt.GptService;
import com.star.easydoc.service.translator.TranslatorService;
import com.star.easydoc.view.inner.SupportView;
import org.jetbrains.annotations.NotNull;

/**
 * 应用程序激活监听器
 *
 * @author wangchao
 * @date 2022/03/13
 */
public class AppActivationListener implements ApplicationActivationListener {
    private static final Logger LOGGER = Logger.getInstance(AppActivationListener.class);

    /** 锁判断 */
    private volatile boolean isActivate = false;

    /** 锁 */
    private static final Object LOCK = new Object();

    @Override
    public synchronized void applicationActivated(@NotNull IdeFrame ideFrame) {
        activate();
    }

    /**
     * 激活
     */
    public void activate() {
        synchronized (LOCK) {
            if (isActivate) {
                return;
            }
            support();
            serviceInit();
            isActivate = true;
        }
    }

    /**
     * 支持
     */
    private void support() {
        AnAction starAction = new NotificationAction("⭐ 去点star") {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e, @NotNull Notification notification) {
                try {
                    Desktop dp = Desktop.getDesktop();
                    if (dp.isSupported(Desktop.Action.BROWSE)) {
                        dp.browse(URI.create("https://github.com/starcwang/easy_javadoc"));
                    }
                } catch (Exception ex) {
                    LOGGER.error("open url failed:https://github.com/starcwang/easy_javadoc", ex);
                }
            }
        };

        AnAction reviewsAction = new NotificationAction("\uD83D\uDC4D 五星好评") {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e, @NotNull Notification notification) {
                try {
                    Desktop dp = Desktop.getDesktop();
                    if (dp.isSupported(Desktop.Action.BROWSE)) {
                        dp.browse(URI.create("https://plugins.jetbrains.com/plugin/12977-easy-javadoc/reviews"));
                    }
                } catch (Exception ex) {
                    LOGGER.error("open url failed:https://plugins.jetbrains.com/plugin/12977-easy-javadoc/reviews", ex);
                }
            }
        };

        AnAction payAction = new NotificationAction("\uD83C\uDF57 加个鸡腿") {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e, @NotNull Notification notification) {
                SupportView supportView = new SupportView();
                supportView.show();
            }
        };

        NotificationUtil.notify("关于EasyJavadoc快捷键",
            "EasyJavadoc默认的快捷键是command(ctrl)+\\,和新版IDEA的AI Assistant插件冲突,请自行修改快捷键");
        NotificationUtil.notify("支持EasyJavadoc", "如果这款小而美的插件为您节约了不少时间，请支持一下开发者！",
            starAction, reviewsAction, payAction);
    }

    /**
     * 服务初始化
     */
    private void serviceInit() {
        EasyDocConfig config = ServiceManager.getService(EasyDocConfigComponent.class).getState();

        TranslatorService translatorService = ServiceManager.getService(TranslatorService.class);
        translatorService.init(config);
        GptService gptService = ServiceManager.getService(GptService.class);
        gptService.init(config);
    }

    @Override
    public void applicationDeactivated(@NotNull IdeFrame ideFrame) {
        applicationActivated(ideFrame);
    }
}

File: easydoc-kdoc/src/main/kotlin/com/star/easydoc/kdoc/listener/AppActivationListener.kt
package com.star.easydoc.kdoc.listener

import com.intellij.icons.AllIcons
import com.intellij.notification.*
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationActivationListener
import com.intellij.openapi.components.ServiceManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.wm.IdeFrame
import com.star.easydoc.kdoc.config.EasyKdocConfigComponent
import com.star.easydoc.kdoc.view.inner.SupportView
import com.star.easydoc.service.translator.TranslatorService
import java.awt.Desktop
import java.net.URI

/**
 * 应用程序激活监听器
 *
 * @author wangchao
 * @date 2022/03/13
 */
class AppActivationListener : ApplicationActivationListener {
    /** 上一次通知时间  */
    @Volatile
    private var lastNoticeTime = 0L

    @Synchronized
    override fun applicationActivated(ideFrame: IdeFrame) {
        support()
        serviceInit()
    }

    /**
     * 支持
     */
    private fun support() {
        if (System.currentTimeMillis() - lastNoticeTime < INTERVAL) {
            return
        }
        val group = NotificationGroup(
            "Easy Kdoc", NotificationDisplayType.BALLOON, true, null,
            AllIcons.General.AddJdk
        )
        val notification = group.createNotification(
            "支持EasyKdoc", "如果这款小而美的插件为您节约了不少时间，请支持一下开发者！",
            NotificationType.INFORMATION, NotificationListener.URL_OPENING_LISTENER
        )

        // 去点star
        notification.addAction(object : NotificationAction("✨ 去点star") {
            override fun actionPerformed(e: AnActionEvent, notification: Notification) {
                try {
                    val dp = Desktop.getDesktop()
                    if (dp.isSupported(Desktop.Action.BROWSE)) {
                        dp.browse(URI.create("https://github.com/starcwang/easy_javadoc"))
                    }
                } catch (ex: Exception) {
                    LOGGER.error("打开链接失败:https://github.com/starcwang/easy_javadoc", ex)
                }
            }
        })

        // 支付
        notification.addAction(object : NotificationAction("☕ 请喝咖啡") {
            override fun actionPerformed(e: AnActionEvent, notification: Notification) {
                val supportView = SupportView()
                supportView.show()
            }
        })
        notification.notify(null)
        lastNoticeTime = System.currentTimeMillis()
    }

    /**
     * 服务初始化
     */
    private fun serviceInit() {
        val config = ServiceManager.getService(EasyKdocConfigComponent::class.java).state
        val translatorService = ServiceManager.getService(TranslatorService::class.java)
        translatorService.init(config)
    }

    override fun applicationDeactivated(ideFrame: IdeFrame) {
        applicationActivated(ideFrame)
    }

    companion object {
        private val LOGGER = Logger.getInstance(AppActivationListener::class.java)

        /** 通知时间间隔  */
        private const val INTERVAL = 7 * 24 * 60 * 60 * 1000L
    }
}